harness/determined/pytorch/_pytorch_context.py

```diff
@@ -467,41 +532,63 @@ def clip_grads(params):
                 stepping the optimizer. If false, you need to call ``optimizer.zero_grad()``
                 manually. Note that if :ref:`optimizations.aggregation_frequency
                 <config-aggregation-frequency>` is greater than 1, ``auto_zero_grads`` must be true.
+            scaler(``torch.cuda.amp.GradScaler``, optional): The scale to use for stepping the
+                optimizer. This should be unset if not using AMP, and is unnecessary if
+                ``PyTorchContext.wrap_scaler`` was called with ``automatic = True``.
         """
 
         check.true(
             auto_zero_grads or self.hvd_config.aggregation_frequency > 1,
             "if optimizations.aggregation_frequency is larger than 1, "
             "you can only set auto_zero_grads to be true. ",
         )
-        if self._should_communicate_and_update():
-            # Communication needs to be synchronized so that is completed
-            # before we apply gradient clipping and `step()`.
-            if self.hvd_config.use and not self._use_amp:
-                optimizer.synchronize()
-
-            parameters = (
-                [p for group in optimizer.param_groups for p in group.get("params", [])]
-                if not self._use_amp
-                else apex.amp.master_params(optimizer)
+
+        if not self._should_communicate_and_update():
+            return
+
+        # Communication needs to be synchronized so that is completed
+        # before we apply gradient clipping and `step()`. In the case of APEX
+        # this is called in backward() instead, so that it's inside the context
+        # manager and before unscaling.
+        if self.hvd_config.use and not self._use_apex:
+            optimizer.synchronize()  # type: ignore
+
+        parameters = (
+            [p for group in optimizer.param_groups for p in group.get("params", [])]
+            if not self._use_apex
+            else apex.amp.master_params(optimizer)
+        )
+
+        if self.hvd_config.average_aggregated_gradients:
+            self._average_gradients(
+                parameters=parameters, divisor=self.hvd_config.aggregation_frequency
             )
 
-            if self.hvd_config.average_aggregated_gradients:
-                self._average_gradients(
-                    parameters=parameters, divisor=self.hvd_config.aggregation_frequency
-                )
+        if clip_grads is not None:
+            if self._scaler and self._auto_amp:
+                self._scaler.unscale()
```

Comment:
Shouldn't this be `self._scaler.unscale_(optimizer)`?

Reply:
Yeah :) I've fixed this twice now and gotten it mixed up in my testing - copying this file back and forth between experiments :D